### PR TITLE
Fortran Intrinsics

### PR DESCRIPTION
**Pull request by urbanjost**
Fortran Intrinsics descriptions have been discussed as an addition to fortran-lang.org.  Herein is a skeleton for the descriptions as a series of MINIBOOK entries.

Preview available at https://fortran-lang.org/pr/347/

### CODE REVIEW

**Comment by milancurcic** [comment]
#build_preview

**Comment by certik** [comment]
Nice! This looks awesome. Did you write all that? I am asking to ensure we don't run into copyright issues.

I noticed sometimes the formatting needs improvements, e.g. here:

https://fortran-lang.org/pr/347/learn/intrinsics/ATAN2

the part `-PI <= atan(x) <= PI.` is inconsistently bold. But we can fix that after merging.

**Comment by urbanjost** [comment]
I started off writing my own for each intrinsic but at one point to have a starting point for each I filled it in with the pages from the Fortran Wiki, which are under the GFDL, which as I understand it allows for them to be modified and reproduced like this. Many have been changed so much "modified" is an understatement. Ones I know I wrote from scratch I marked at MIT. Many more are from scratch but I didn't have anything marking a lot of them as such so to be conservative I marked them as GFDL as they might have started from the Wiki versions. Since all descriptions of the intrinsics in some way depend on the standard definition of Fortran I have puzzled over exactly how anyone can produce Fortran documentation that describes Fortran without issues with the copyright of the standard. The intent here is that the community as a whole can write new ones for each GFPL-licensed minibook and generate an MIT-only based collection. Each of the vendors has produced a description of the intrinsics, often identical or nearly identical to the descriptions in the Fortran standard, which bolstered my interpretation, but I could be wrong. There is a section in the introduction that is also the second index item that I meant as explanation. Otherwise, it would have to be torn down to just an index and volunteers could be asked to write descriptions (from memory?). Perhaps a vendor has an easier time as they are describing their implementation (?). I started each one as plain text and used some simple scripts to initially make them markdown, and will go through and correct the results (have already been doing this) but learning some of the vagaries of this version of markdown partly by looking at the results (and already have been rewriting some of the results) but there are quite a few unexpected ones such as a / needing escaped and such that can be easily corrected; but already spent too much time on this if it cannot proceed so wanted to present what I have so far and see if we can all agree on going forward, or coming up with an alternate plan. Since you are working on LFortran I was thinking you might have some quick answers on exactly how you can document your results if they are based on the Fortran Standard. I thought the text in the Intro explained the pedigree, but perhaps not well enough.

**Comment by urbanjost** [comment]
For some reason the links in the text are not working, but worked on the fork I made I was trying this with.

**Comment by certik** [comment]
@urbanjost thanks for the feedback. I wish FortranWiki was licensed under MIT or some of the permissive Creative Commons licenses (I can't remember which one it is). 

I think the best way forward is to simply note for each part where it is taken from. If you wrote it, then it would be MIT, which is the default, so you don't have to mark it. If it comes from FortranWiki, then credit them and mark GDFL (hopefully it doesn't mean we have to license the whole website under GDFL?). And stuff that came from the standard, let's reference that it came from the standard.

That should allow us to move forward. 

I would love to reuse all this for LFortran eventually, so that the compiler can point to the definition, say for error messages, as well as bundle the documentation in some way so that things are available locally, so that it works without internet access. Perhaps we can integrate with VSCode somehow, I don't know yet.

Yes, it would be nice to figure out how to handle the GDFL parts as well as parts copied from the standard, so that we are legally on a solid ground. I don't have the answer right now, but I think the above approach should allow us to get moving, and we can figure out the answer as we go.

**Comment by urbanjost** [comment]
It was written as part of fpm, and later as the fpm-man plugin. The original intent was to make it available from the command line.
So that part might be easy. I can try converting back to text and then to Fortran statements and see how it looks. I just installed pandoc, which it says reads github markdown and can output ASCII text.  If that works it would take a minor rework to make it automated to turn this into a fpm-man replacement.

The Fortran standard says you can reproduce parts of it by permission, with a possible fee. That might be the ultimate permission or the ultimate roadblock. 

I used to use a nice error message utility called explain. Every code included a message routine that took a product name and an error code.  If you hit an error an external library was called and given the two codes. It then produced the message using a message catalog.  This allowed for easy multi-lingual capabilities as you could have a message catalog for any supported language, or even add local info to the messages, which were kept in ASCII. It had a terse and verbose mode. When called from programs the default was terse, when called manually the default was verbose.  So if you got a message like 
   f90 #554 - improper something-or-other

you could then enter "explain f90 544" and get a full description of the error.  You could also turn off warnings or change their severity level, so if you turned on all the warnings the compiler (or other product) generated, and then realized every line was telling you  "f90 #300 FORTRAN Statements must be all uppercase" or "f90 #200 Variable names are limited to six characters" you could turn those off with an environment variable.  Except for occasional problems with the environment variable that was used to give a list of directories to search for the catalogs it was great.

For security reasons some programs cleared the environment table on startup, which caused the programs to not find the catalogs, and once one of the developers accidently let his personal message catalog get sent to a customer and during a demo in front of the corporate executives of the customer an error produced a barage of profanity; which made quite a few people laugh but outraged some other rather important ones.  Other than that it was great.

**Comment by certik** [comment]
Haha, great story. Yes, indeed we need translations and explanations for each error in LFortran, along the lines you wrote. Rust does something similar.

**Comment by awvwgk** [comment]
Thanks, great work. I have one stylistic remark, could we get away from the all caps keywords and headers?

**Comment by urbanjost** [comment]
Went to \#|#\#\# to de-emphasize the headers, but the man-style is being used (which requires the all-cap headers) not only because it is a good style many are familiar with but because I am hoping these pages can be automatically rendered into flat text for use in providing data for a simple in-line tool that can be written in Fortran (ie. that the data can be used for fpm-man) and to create actual manpages.  Experimenting with pandoc(1) for that and it looks very feasible. Perhaps getting triple use out of the descriptions via an automated process is asking for too much, as it limits you to the ASCII character set for the most part, making representing more complex mathematics and probably some other things too much;  but hoping to get this in place not just for the intrinsics but for the stdlib procedures too, where part of the process of providing a routine in stdlib would be to create a description here as well.  At least on my screen going to a smaller size helped, I think.   Maybe should have stuck with HTML but the Markdown seemed like the write way to go, as it has a lower entry barrier so a lot more people would feel comfortable generating the documention, I think.

**Comment by urbanjost** [comment]
pandoc worked reasonably well considering I do not know what all the switches do yet.
If you change the name of the directory where the source markdown is on a Linux or Cygwin or WSL
environment and have pandoc installed it makes reasonable man-pages automatically, maybe (text needs a little tweeking, perhaps). I find that enticing, but a program like fman/fpm-man could provide similar functionality and not require following the man-page syntax nor the existence of man(1) for that matter.




```bash
#!/bin/bash
header(){
cat <<\EOF
." Text automatically generated by txt2man
.TH "abs" "3" "November 20, 2021" "" "" " "
." -----------------------------------------------------------------
." * set default formatting
." disable hyphenation
.nh
." disable justification (adjust text to left margin only)
.ad l
." -----------------------------------------------------------------
EOF
}
cd $HOME/github/FORK/fortran-lang.org/learn/intrinsics
mkdir -p /tmp/man/man3
for NAME in *.md
do
   SHORTNAME=$(basename $NAME .md)
   SHORTNAME=${SHORTNAME,,}
   echo "NAME: $NAME to $SHORTNAME"
   (
      header
      sed -n -e '\%^### NAME%,${ p }'  $NAME|
         pandoc -f gfm -t man --columns=72 
   )>/tmp/man/man3/$SHORTNAME.3fortran
done
cd /tmp/man
mandb -c .
env MANWIDTH=256 man --manpath=/tmp/man -k .|col -b
env MANWIDTH=80  man --manpath=/tmp/man --regex '.*'|col -b
exit
```

**Comment by awvwgk** [comment]
Thanks, let's update the #build_preview.

**Comment by awvwgk** [comment]
Ugh, our static side generator doesn't seem to scale well, 1m 11s for the current default build, with the new minibook we are at  3m 43s. We can still stand this build time, but that might not stay sustainable in the long run for more learning resources. Not relevant for this PR but something to keep in mind for the future.

**Comment by certik** [comment]
I have good experience with Hugo. Very fast builds.

**Comment by awvwgk** [comment]
#build_preview

**Review by awvwgk — APPROVED** [review]
Really amazing work, John @urbanjost. Looking forward to have this available at the webpage.

**Comment by urbanjost** [comment]
Unfortunately a long ways to go, and I want to add basic statements as well  (EXIT, STOP, ALLOCATE ....) and link to other resources. Did not seem to get much interest on Fortran Discourse either.  Waiting for a response from ISO committee on permission to use excerpts from the standard. Hope that does not backfire.

**Comment by certik** [comment]
@urbanjost I thiought you got great response. We all want this!

**Comment by urbanjost** [comment]
I was looking more for feedback from others that might want to contribute and was looking to use that to make an easier way (probably allowing for plain text contributions) as there are quite a few requirements for contributing directly to the PR that I thought might be off-putting.    As a lot of Fortran developers are largely self-taught and there are less freely available resources available (at least historically, it is getting better) I think starting with expanded descriptions of the intrinsics will be useful, but is going to be slow going if only a few people contribute; and I honestly have yet to use the atomic procedures although I am familiar with them. Coming up with some good examples for some of the routines like that is going to be time-consuming.

When I build this on my machine all the links work, but when using the build on the pull request the automatically built alphabetical links work, but the ones in the body of the index do not.  Do those work for anyone?

**Comment by milancurcic** [comment]
> When I build this on my machine all the links work, but when using the build on the pull request the automatically built alphabetical links work, but the ones in the body of the index do not. Do those work for anyone?

Same here, they work in my local build.

**Comment by arjenmarkus** [comment]
I'd love to help out with this, but I am not sure how I can contribute ;).

I just checked the built tutorial and found that you can nicely scroll in
the list and view the various items, but when you do click on one, the list
immediately scrolls to the beginning. Not enitirely what I expected.

Op do 2 dec. 2021 om 03:12 schreef urbanjost ***@***.***>:

**Comment by arjenmarkus** [comment]
Also note: a number of entries are doubled - acosh, co_min, get_command,
ibclr, log_gamma, (after not an empty entry), rank, selected_int_kind

Op do 2 dec. 2021 om 11:44 schreef Arjen Markus ***@***.***>:

**Comment by urbanjost** [comment]
Ends up if you make any typo at all the index repeats the last value it did get, I guess. And apparently the title "null" is a reserved word.  It has not been rebuilt and tried but I think the duplicates are all gone now; but had to stick a space in "null" for the time being; tried decimal codes and that did not help. So that already helped.

So the index is automatically built so I do not have a lot of control over it that I know of; the RHS index is written by hand but the links seem to work fine on my local machine but do not work on fortran lang; so I stuck several variants of the URL for the MERGE function in so the next time it gets built I can see if any of them work.

New to the MINIBOOK format and how to build it locally myself; basically just followed the site download directions on the github site so not sure what the optimal way is for everyone to contribute that way;  but just plain-text or markdown versions of any of the intrinsics, especially foir the ones marked with the GNU license in the index; hoping to rewrite those and make MIT-licensed versions. 

Need to come up with some better process but so far just struggling with getting the skeleton laid out myself. Ideas welcome.

File: learn/intrinsics/NULL.md
Comment (awvwgk) [inline learn/intrinsics/NULL.md:3]:
```suggestion
title: "null"
```

**Comment by urbanjost** [comment]
Thought I tried the quotes (I tried Unicode decimal and hex, ... too) but it works. I might quote all the titles. 
Thanks!   Still have odd things like the links on the RHS of the index.html page not working when on fortran-lang.org but working on my local server, and not being able to scroll to the bottom of the LHS index in Firefox but the basic structure looks to be coming together OK.

**Comment by urbanjost** [comment]
If I could ask for a build, I need a trail build to see what links work on a build on fortran-lang.org; I have asked for permission and
clarification on use of the ISO standard,  but I think based on this document that the ISO standard can be used as a reference
in this context so this can go forward, albeit this a lot to go this will create a complete skeleton for the standard generic intrinsics
that hopefully everyone can help complete.

https://www.iso.org/files/live/sites/isoorg/files/store/en/PUB100206.pdf
>
>   Copy parts of a standard for your book or software
>   
>   Citing standards or including extracts of standards is encouraged as
>   long as there is the correct acknowledgement and the conditions of your
>   licence are respected. Please contact IEC or ISO for authorization.

**Comment by awvwgk** [comment]
> If I could ask for a build

You can just add a comment with the phrase #</space>build_preview and you will get a new preview. Give it a try.

**Comment by urbanjost** [comment]
#build preview

**Comment by urbanjost** [comment]
#178

**Comment by urbanjost** [comment]
#build preview

**Comment by awvwgk** [comment]
```
#build preview
      ^
```
You are missing an underscore, try with #</space>build_preview.

**Comment by urbanjost** [comment]
So, not sure what you mean, apparently. If I fill out a comment field such as this with the line "#build preview" nothing happens that I can see except it keeps trying to change the string to a PR number. Not familiar with this feature or where to look for information on it. Exactly how do I trigger a rebuild?

**Comment by awvwgk** [comment]
#build_preview

**Comment by awvwgk** [comment]
There is a workflow running scanning all comments posted for the string #</space>build_preview, if it occurs in any context it will trigger a new preview build. Note that this comment and my previous ones did not trigger the preview build because I put a non-printing character in via `#</space>build_preview`.

The difference between your previous comments and mine above is the underscore in #</space>build_preview. Using # as character to signal a special command is a rather suboptimal design choice, since the GitHub UI will try to expand it to a reference. Also, I'm not sure if the preview build workflow is actually documented anywhere.

Another thing about the current mechanism is that it is kind of hard to explain without either triggering a new preview build or causing confusion...

**Comment by urbanjost** [comment]
#build_preview

**Comment by urbanjost** [comment]
#build_preview

The basic layout is working including the links and a general style has emerged that I think works; except for an oddity where I have to reduce my zoom level for the entire auto-built index to appear. There is still a large amount of work on the descriptions themselves to be done, which I would encourage everyone to help with; with good example programs needed and/or links to additional resources but this is at the point of being usable in my opinion, and still can be converted to plain text and actual man-pages via some basic GNU scripting and the use of pandoc(1); so in this form it can still be used to generate ULS documentation and be used as input to create an fpm-man(1) utility for access via a CLI tool; although the addition of mathematical expressions and links and graphics would complicate that. Have not tried it, but pandoc(1) can generate HTML and Adobe PDF as well, but even if not can use groff(1) on the man-pages and/or JavaScript as is done on the M_intrinsics repository to create a single document as well, hopefully. So how far to go beyond ASCII documentation at the risk of loosing the ASCII interface alternative is a big question.  It will be an extended effort to go significantly further than this so I am looking for feedback on how well the MINIBOOK format appears to work to everyone, etc.

**Comment by milancurcic** [comment]
The top-level table in the latest preview looks very nice, great work @urbanjost!

**Comment by urbanjost** [comment]
#build_preview

Thanks. Well, if you build it will they come?

**Comment by urbanjost** [comment]
SYSTEM_index.md did not seem to rebuild

#build_preview

**Comment by urbanjost** [comment]
#build_preview for EPSILON draft

**Comment by urbanjost** [comment]
#build_preview  

Too fast, or should not put comment on same line? Try again

**Comment by awvwgk** [comment]
> Too fast, or should not put comment on same line? Try again

I think the complete build takes at least 3 to 4 min at the moment. For quick checks building locally might be an option as the rebuild is incremental, but at least for me that stopped working after I upgraded to Ruby 3 on my system.

**Comment by urbanjost** [comment]
Ahhh. I went back to ruby 2.7 and I can do local builds again; plus I was not waiting long enough. As soon as I got a response I was looking, but if I just wait a few minutes after it shows up. Have to take a break on this but wanted to get some of the smaller things out of the way so this was at a presentable state. It is at a point where everything works and the response has been mild so I will work on this locally again and experiment a bit more with mathematical expressions and graphics unless something changes (never used kramdown before, etc...). Thanks again!

#build_preview

**Comment by urbanjost** [comment]
#build_preview

**Comment by urbanjost** [comment]
#build_preview

for EXP

**Comment by urbanjost** [comment]
#build_preview

**Comment by urbanjost** [comment]
#build_preview

LEN

**Comment by urbanjost** [comment]
#build_preview

Add an example of specifying the length of a _character_ variable at run time even if not allocatable.

**Comment by urbanjost** [comment]
#build_preview

**Comment by urbanjost** [comment]
#build_preview

errata and formatting corrections

**Comment by urbanjost** [comment]
#build-refresh

**Comment by urbanjost** [comment]
#build-rebuild

Add copies of preliminary man-pages and fman(1) program

**Comment by awvwgk** [comment]
I'm not sure if we can properly host Fortran source on the webpage. In any case a separate repository with the source and a release tag holding any generated artifacts, which are linked from the webpage seem preferable.

**Comment by urbanjost** [comment]
#build-preview
moved supplemental files to AlterVista

**Comment by urbanjost** [comment]
#build_preview

**Comment by awvwgk** [comment]
> moved supplemental files to AlterVista

If you like we could maintain them in a repository in the @fortran-lang namespace.

cc @fortran-lang/admins

**Comment by urbanjost** [comment]
If they cannot be allowed in the directories themselves I am fine with them being on an ID of mine; it is experimental at this time.
Not quite sure what it would take to have them under fortran-lang.org but right now they change so frequently it is actually convenient with them on AlterVista, at least for me.

#build_preview

**Comment by urbanjost** [comment]
#build_preview

**Comment by urbanjost** [comment]
I was hoping this would be accessible from at least GNU/Linux and Unix via programs like lynx/links/w3m; but the index shows on each page, which makes it awkward:
```bash
    lynx https://fortran-lang.org/pr/347/learn/intrinsics/ACHAR
```
perhaps I can directly insert HTML to mask that, but I do not think that is supported by the CLI browsers.  Any thoughts appreciated. I have not found a way to turn off the index on the individual pages as they are built automatically, so I do not see an easy way to insert directives into the index, other than perhaps creating everything as HTML instead of as markdown.

#build_preview

**Comment by urbanjost** [comment]
#build_preview

**Comment by urbanjost** [comment]
#build_preview

**Comment by urbanjost** [comment]
#build_preview

VERIFY and slidy view of procedures. I think I am becoming a slidy fan.

**Comment by urbanjost** [comment]
#build_preview

**Comment by urbanjost** [comment]
#build preview

**Comment by urbanjost** [comment]
#build preview

The master files are the kramdown files, so anyone else that wants to contribute can just change the files. A particular style that is basically a man-page format is being followed, with a simple set of rules being followed to allow for simple parsing and to maintain compatibility with pandoc(1) for conversions. As an alternate way of contributing, particularly example programs, see
https://urbanjost.github.io/fortran-intrinsic-descriptions/;  or we can move this to the Fortran Wiki, which has a much lower bar for contributing and less issues with potential copyright issues.  I have not gotten the type of definitive answer from the ISO standard support yet, although there is a page saying that as long as you properly acknowledge the standard as a referenence that reasonable use is permitted. Anyone published a Fortran manual that has some feedback on how to navigate properly using the standard without infringing on it?

**Comment by awvwgk** [comment]
> I have yet to get a LaTex formula to work with kramdown on this page. Anyone have an example?

I remember using MathJax with jekyll / kramdown in the past. There was an issue with TeX and markdown sharing a lot of control characters, something hacky was needed to get those working together like passthroughs or similar. I could look it up, but maybe there is a better solution than TeX? I recall @certik commented on the usage of SymPy for rendering formulas instead of TeX.

**Review by milancurcic — APPROVED** [review]

**Comment by milancurcic** [comment]
@awvwgk @henilp105 @certik 

In light of the new GSoC project by @henilp105 and future transition to Sphinx+MyST parser, should we go ahead and merge this? I think so--we'd have a good intrinsics reference on the website (despite longer builds), and a good baseline for transition to Sphinx. The intrinsics reference files are in Markdown, which I think is a good format to use as the "source" for translation to html.

**Comment by urbanjost** [comment]
Have not quite sorted out Sphinx, but the current build converts them to HTML, and pandoc converts to reStructuredText, with something like
```bash
pandoc -f markdown_mmd -t rst -i MAXLOC.md  -o ../RST/MAXLOC.md
```
so (theoretically, have not tried it) they should convert to reStructuredText (maybe with a little massaging with sed(1) or something like it); which I believe is what Sphinx uses; if that is of any value.

**Comment by henilp105** [comment]
@awvwgk @milancurcic 
It would be easy to implement/migrate the intrinsics part to Sphinx as it is already written in markdown, We can easily parse it using Sphinx+MyST parser and they will be directly converted to html and Transitioning to Sphinx+Myst parser would also considerably decrease the Build times.

Thanks and Regards,
Henil Shalin Panchal

**Comment by milancurcic** [comment]
@certik can this PR be merged?

**Comment by certik** [comment]
I think so.

**Comment by milancurcic** [comment]
From a cursory look at the preview, nothing stands out as broken, so I'll go ahead and merge this.

Thanks a lot @urbanjost for this monumental effort and contribution, and thanks to all reviewers.

**Pull request merged**
